README.md
# Screen Shot Plugin

This is a JBrowse plugin
 
This plugin allows the user to take a high quality screenshot of the browser.
This plugin leverages the power of PhantomJS to create an image of the browser. Unfortunately, images are created using only the backend track settings so changes made within the browser interface will not included in the screenshot.

## Special Thanks
This plugin would not be possible without [phantomJS](http://phantomjs.org/), specifically [PhantomJS Cloud](https://phantomjscloud.com/), which provides phantomJS as a headless software-as-a-service. 


## Install

For JBrowse 1.11.6+ in the _JBrowse/plugins_ folder, type:  
`git clone https://github.com/bhofmei/jbplugin-screenshot.git ScreenShotPlugin`

**or**

downloaded the latest release version at [releases](https://github.com/bhofmei/jbplugin-screenshot/releases).  
Unzip the downloaded folder, place in _JBrowse/plugins_, and rename the folder _ScreenShotPlugin_

## Activate

Add this to _jbrowse.conf_ under `[GENERAL]`:
```
    [plugins.ScreenShotPlugin]
    location = plugins/ScreenShotPlugin
```

If that doesn't work, add this to _jbrowse_conf.json_:
```
    "plugins" : {
        "ScreenShotPlugin" : { "location" : "plugins/ScreenShotPlugin" }
}
```

## PhantomJS Cloud

### Fine print
Using this plugin means you agree to PhantomJS Cloud's [Terms of Service](https://phantomjscloud.com/tos.html).
The important points
1. Only have one PhantomJS Clound account.
2. PhantomJS cloud is provided 'as-is' and does not guarantee service at all times.

### Accounts
This plugin will work directly out of the box with PhantomJS Cloud. PhantomJS Clous offers a "demo" version of its services where each IP address is limited to 100 requests per day.
If this is insufficient, you can create a free account with PhantomJS Cloud which allows 500 requests per day to that account. See [PhantomJS Cloud](https://phantomjscloud.com/pricing.html) for more information.

If you do have your own account with PhantomJS, include your user API key when activating the plugin.

For _jbrowse.conf_
```
    [plugins.ScreenShotPlugin]
    location = plugins/ScreenShotPlugin
    apiKey = <insert_api_key>
```
For _jbrowse_conf.json_
```
    "plugins" : {
        "ScreenShotPlugin" : { "location" : "plugins/ScreenShotPlugin",
                            "apiKey": "<insert_api_key>" }
    }
```

You can also use your own API key by inputting it in the dialog pane with the "Use Custom ApiKey" parameter.
    
## Use
Click the "Screen shot" button in the browser. A dialog box will open with options for the screenshot. You can also press the `s` key as a shortcut to open the dialog box.

When increasing the zoom factor, you will likely need to increase the height and width to ensure a full image.

When selecting the PDF output, you must choose a page size (A3, A4, A5, letter, legal, or tabloid) and orientation (landscape or portrait). 
The height and width parameters do not specify output page size. Now they specificy the "viewport" (portion of screen to be printed) of the image to be placed on the output page. This may require some trial and error to get the best view.  
Also, the tracklist does not always display well when using PDF output. When selecting the PDF output option, it will automatically uncheck "Show tracklist".

Hiding track labels only works for JBrowse v1.12.3 when HideTracksButton plugin is activated. If browser is an earlier version and/or the plugin is not activates, track labels will always be shown.

**Note:** Due to the nature of URL-encoded screenshots, default track configurations will be used unless overriden by settings selected in this dialog box. Locally added tracks, such as combination tracks, will not be included.

### HTML and SVG Exporting
Canvas Feature tracks are great when browsing but don't show up as well in screenshots. 
If you'd like to convert a canvas feature track to HTML feature track when taking the screenshot, click the "HTML feature" checkbox on that track's configuration.
The style/look of the features is controlled by the track "style.className" configuration.

HTML/SVG track exports are available for Canvas Feature tracks and Alignment2 tracks by default.
Additional support for XYPlots and Density plots are available with an [additional plugin](#wiggle-svg-plot-plugin).

The default configuration is to leave each as-is and the HTML/SVG-style option must be selected for each support track.
To change this default behavior (i.e. all tracks that can be converted to HTML/SVG are converted unless the option is deselected for each track), change the configuration when activating the plugin.

In _jbrowse.conf_,

```
[plugins.ScreenShotPlugin]
location = ...
htmlFeatures = true
```


## Support for Additional Plugins
### Methylation Plugin
This plugin includes support for the [MethylationPlugin](https://github.com/bhofmei/jbplugin-methylation).  
Make sure the plugin ID<sup>1</sup>  is `MethylationPlugin` and the plugin is activated.

For MethylationPlugin v3.1.0+, using HTML-based methylation track in the screenshot is supported

### Small RNA Plugin
This plugin includes support for the [SmallRNAPlugin](https://github.com/bhofmei/jbplugin-smallrna).  
Make sure the plugin ID<sup>1</sup> is `SmallRNAPlugin` and the plugin is activate. 

For SmallRNAPlugin v1.4.0+, using HTML-based small rna track in the screenshot is supported.

### Stranded XYPlot Plugin
This plugin includes support for the [StrandedPlotPlugin](https://github.com/bhofmei/jbplugin-strandedplot).  
Make sure the plugin ID<sup>1</sup> is `StrandedPlotPlugin`and the plugin is activated. 

For StrandedPlotPlugin v1.1.0+, using SVG-based tracks in the screenshot is supported.

### Nucleotide Density Plugin
Nucleotide Density Plugin was rebranded to [MotifDensityPlugin](https://github.com/bhofmei/jbplugin-motifdens). Download and use MotifDensityPlugin instead.

Nucleotide Density tracks are still available for screenshots as a normal density track, however SVG-based screenshot tracks are no longer supported.

### Motif Density Plugin
This plugin includes support for the [MotifDensityPlugin](https://github.com/bhofmei/jbplugin-motifdens).  
Make sure the plugin ID<sup>1</sup> is `MotifDensityPlugin`and the plugin is activated. 

For MotifDensityPlugin v2.0.0+, using SVG-based tracks in the screenshot is supported.

### Wiggle SVG Plot Plugin
For XYPlot and Density plots to be exported as SVG-based tracks, download and activate the [Wiggle SVG Plot Plugin](https://github.com/bhofmei/jbplugin-wigglesvg).
Make sure the plugin ID<sup>1</sup> is `WiggleSVGPlotPlugin` and the plugin is activated.


### SeqViewsPlugin
This plugin only needs to know if the [SeqViewsPlugin](https://github.com/bhofmei/jbplugin-seqview) is activated.

It will automatically look for a plugin with the plugin ID `SeqViewsPlugin`. If the SeqViews Plugin has been activate but with a different plugin ID, indicate that `seqViewsPlugin = true`.

In _jbrowse.conf_,
```
    [plugins.ScreenShotPlugin]
    ...
    seqViewsPlugin = true
```

Optionally, you can specify `seqViewsPlugin = false` if you do not want to include support for it.

<sup>1</sup>For jbrowse.conf, the plugin ID is found as `[plugins.ID]` for each plugin.  
In jbrowse_conf.json, the plugin ID is found as `"plugins":{"ID":{"location":"..."}}`

### Other plugins
Support for other plugins can be added based on demand. To request support for other plugins, either

a. Create an [issue](https://github.com/bhofmei/jbplugin-screenshot/issues) in GitHub. Include the plugin name and URL as well as what configuration options should be supported

or

b. Fork this plugin, add the support, then submit a pull request.

## Issues/Bugs
This plugin is difficult to test and debug because it relies on PhantomJS Cloud.

### General troubleshooting
1. **If account login information is required to access the browser, it will not work.** There are ways around this, although they are not currently implemented because I haven't figured out a good way to do it securely and lack of demand.
2. **If the PhantomJS window is blank**
  * In the full request URL, change the render type to "json" and resubmit. Look for error messages.
  * Check is the request URL. In the full URL sent to PhantomJS cloud, extract the portion at `.../?request:{url:"get_this_url",...}`. Replace `%26` with `&` and `%3A` with `:` and open this updated link. Look for any error messages in the console.
3. If the PhantomJS window gives an error message about being out of credits, create a PhantomJS account.

### Extra help
For additional debugging help, submit an [issue](https://github.com/bhofmei/jbplugin-screenshot/issues). Include the URL when submitting an issue.

## Future Improvements
- Show/hide clip marker
- Remember configuration settings after closing the dialog


CHANGELOG.md
# Change Log

## [Unreleased]

## [v1.6.10] - 2017-09-26
- ADDED support for extended modifications (4mC, 5hmC, 6mA) for methylation plugin

## [v1.6.9] - 2017-08-28
- DELETED support for NucleotideDensityPlugin since it was rebranded as MotifDensityPlugin
- ADDED support for MotifDensityPlugin

## [v1.6.8] - 2017-08-22
- ADDED Nucleotide density plugin SVG support

## [v1.6.7] - 2017-08-18
- bug fixes

## [v1.6.6] - 2017-08-17
- Support for SVG-style XYPlot and Density tracks with WiggleSVGPlotPlugin
- Additional plugins must have specified plugin ID
- updated README to include Wiggle SVG Plot plugin and exact plugin id
- Density tracks do not have y-scale option

## [v1.6.5] - 2017-08-16
- minor bug fixes

## [v1.6.4] - 2017-08-11
- ADDED HTML style available or stranded xy tracks
- UPDATED version checks for small rna plugin, methylation plugin, stranded plot plugin to ensure html/svg tracks are included

## [v1.6.3] - 2017-08-08
- ADDED HTML style available for methylation tracks (must have latest version of methylation plugin)

## [v1.6.2] - 2017-08-03
- ADDED HTML style feature options available for small rna alignments, canvas feature tracks, and alignment2 tracks

## [v1.6.1] - 2017-08-02
- FIXED with canvas features/html features option

## [v1.6.0] - 2017-07-28
- Refactor ScreenShotDialog file to separate out parameter logic
- ADDED option to convert CanvasFeature tracks to HTMLFeature tracks
  - this is useful for PDF rendering so annotation objects are movable
- Bug fix for plugin configuration

## [v1.5.0] - 2017-07-27
- ADDED option to use a custom apikey
- useful for getting aroung 100 page/day limit of the default apikey

## [v1.4.1] - 2017-03-16
- ADDED support for SNPCoverage tracks

## [v1.4.0] - 2017-03-16
- ADDED option to allow for extra rendering time
- default max render time is 35 s; can be increased to 5 min (300 s)

## [v1.3.3] - 2017-03-09
- FIXED data undefined issue

## [v1.3.2] - 2016-12-19
- FIXED small rna hide/show in screenshot (previously no small rna's were hidden when taking the screenshot)

## [v1.3.1] - 2016-11-31
- UPDATE for formatting

## [v1.3.0] - 2016-11-30
- ADDED support for small rna plugin (see readme for additional details)

## [v1.2.0] - 2016-11-04
- allows for pdf output option with some pdf specific parameters
- PDF output requires selecting a page size and orientation
- height/width parameters are used for "viewport" not output page size

## [v1.1.1] - 2016-09-12
- FIXXED bug for SeqViewsPlugin configuration

## [v1.1.0] - 2016-09-12
- for canvas-feature type tracks, now can specificy display mode (normal, compact, collapsed)
- when SeqViewsPlugin is activated, can also specificy display style (default, features, histograms)

## [v1.0.5] - 2016-09-08
- FIXED issue with track-specific y-scale radio buttons not correctly checked when dialog is open
- UPDATED check for XYPlots which includes StrandedXYPlots from StrandedPlotPlugin

## [v1.0.4] - 2016-08-30
- ADDED animal methylation coloring

## [v1.0.3] - 2016-08-30
- ADDED methylPlugin configuration parameter to correctly handle MethylationPlugin
- FIXED path to methylation plugin for checkboxes to work properly

## [v1.0.2] - 2016-08-30
- UPDATED dialog so track-specific settings are at the bottom
- track specific setting uses tab container instead of accordian since accordian had problems with more than 8-ish tracks

## [v1.0.1] - 2016-08-29
- FIXED issue with track specific setting where settings for tracks 10+ were ignored

## [v1.0.0] - 2016-08-11
- ADDED track-specific settings

## [v0.0.4] - 2016-08-03
- ADDED render quality as output option
- dialog box is reorganized

## [v0.0.3] - 2016-07-22
- DELETED PDF functionality as it is not functional

## [v0.0.2] - 2016-07-22
- functional with methylation context corrected
- hide track label not supported

## [v0.0.1]
- first functional version

## [v0.0.0] - 2016-07-18
- initial set-up

js/EncodeDecodeUtil.js
define('ScreenShotPlugin/EncodeDecodeUtil', [
    'dojo/_base/declare',
    'dojo/_base/array',
    'dojo/_base/lang',
    'dojo/json'
    ],
  function (
    declare,
    array,
    lang,
    json
  ) {
    var EncodeDecodeUtil;
    EncodeDecodeUtil = {
      encode: function (inputs) {
        // returns string with encode options for screenshot
        var gInputs = inputs.general;
        var tInputs = inputs.tracks;
        //console.log(tInputs);
        return this._encodeGeneralSettings(gInputs) + this._endcodeTrackSettings(tInputs);
      },

      encodePhantomJSSettings: function (params) {
        // params include url, format, height, width, zoom, time
        // ?request={url:"http://www.highcharts.com/demo/pie-donut",renderType:"jpg",renderSettings:{zoomFactor:2,viewport:{width:100,height:500}}}
        // split parameters into pdf type and image type
        var outDict = {
          url: params.url,
          renderType: params.format.value
        };
        var renderDict = {
          zoomFactor: params.zoom.value,
          quality: params.quality.value
        };
        //var outDict = {url: params.url, renderType: params.format.value, renderSettings: {zoomFactor: params.zoom.value, viewport: {width:params.width.value, height: params.height.value}}};
        // check PDF
        if (params.format.value === 'PDF') {
          // page needs split by space, then format, orientation
          var pageFormat = params.pdf.page.value;
          pageFormat = pageFormat.split(' ');
          renderDict['pdfOptions'] = {
            format: pageFormat[0],
            orientation: pageFormat[1],
            footer: null
          };
          renderDict['viewport'] = {
            width: params.pdf.pdfWidth.value,
            height: params.pdf.pdfHeight.value
          };
        } else {
          renderDict['viewport'] = {
            width: params.image.width.value,
            height: params.image.height.value
          };
        }
        // check for extra time
        if (params.time.value) {
          outDict['requestSettings'] = {
            maxWait: (params.time.extra.value * 1000)
          };
        }
        outDict['renderSettings'] = renderDict;
        var outString = json.stringify(outDict);
        outString = outString.replace(/\"([^(\")"]+)\":/g, "$1:");
        return '?request=' + outString;
      },

      decode: function (inStr, tracks) {
        // returns javascript object to be applied
        // split inStr
        var opts = inStr.split('~');
        var trackList = tracks.split(',')
        var gSettings = this._decodeGeneralSettings(opts[0])
        var tSettings = this._decodeTrackSettings(opts.slice(1), trackList);
        //console.log(tSettings);
        return {
          general: gSettings,
          tracks: tSettings
        };
      },

      _encodeGeneralSettings: function (params) {
        // locOver, menu, methylation, nav, trackList, trackSpacing, labels, zoom, small rna
        var output = '';
        var eLabels = {
          zoom: 'z',
          trackSpacing: 'p',
          locOver: 'o',
          trackList: 'r',
          nav: 'n',
          menu: 'u',
          labels: 'b',
          methylation: 'm',
          smallrna: 's'
        };
        var param;
        var thisB = this;
        for (param in params) {
          var data = params[param];
          if (param === 'methylation'){
            var types = ['CG', 'CHG', 'CHH', '4mC', '5hmC', '6mA'];
            output += eLabels[param];
            array.forEach(types, function (t) {
              output += thisB._encodeBoolean(data[t])
            });
          }
          else if (param === 'smallrna') {
            output += eLabels[param];
            var types = ['21', '22', '23', '24', 'pi', 'Others'];
            array.forEach(types, function (t) {
              output += thisB._encodeBoolean(data[t])
            });
          } else if ((param === 'zoom') || (param === 'trackSpacing'))
            output += eLabels[param] + data.value;
          else
            output += eLabels[param] + this._encodeBoolean(data.value);
        }
        return output;
      },

      _endcodeTrackSettings: function (tracks) {
        var output = '';
        // go through object
        var t, params;
        for (t in tracks) {
          params = tracks[t];
          // if we need to encode params
          if (params.hasOwnProperty('opts') === false) {
            output += this._encodeTrack(params);
          }
        }
        return (output);
      },

      _encodeTrack: function (params) {
        // q[0|1] quantitative, y[0|1|2|3] yscale none, center, left, right
        // h# track height, i# min, x# max v[0|1] html features
        // d[0|1|2] display mode normal, compact, collapse
        // f[0|1|2] display style default, features, histograms
        var eLabels = {
          height: 'h',
          min: 'i',
          max: 'x',
          quant: 'q',
          ypos: 'y',
          mode: 'd',
          style: 'f',
          html: 'v'
        };
        var optsDict = {
          ypos: {
            'none': 0,
            'center': 1,
            'left': 2,
            'right': 3
          },
          mode: {
            'normal': 0,
            'compact': 1,
            'collapsed': 2
          },
          style: {
            'default': 0,
            'features': 1,
            'histograms': 2
          }
        };
        var defaultOpts = {
          ypos: 1,
          mode: 0,
          style: 0
        };
        var param, data;

        var output = '~' + params.trackNum;
        // loop through parameters
        for (param in params) {
          data = params[param];
          if (param in {
              'quant': 1
            })
            output += eLabels[param] + this._encodeBoolean(data);
          else if (param in {
              'html': 1
            })
            output += eLabels[param] + this._encodeBoolean(data.value);
          else if (!(data === undefined || data.value === undefined || data.value === null || eLabels.hasOwnProperty(param) === false)) {
            output += eLabels[param]
            // ypos, mode, style
            if (param in {
                ypos: 1,
                mode: 1,
                style: 1
              }) {
              if (optsDict[param][data.value] === undefined){
                console.warn('missing value for ' + param + ', ' + data.value);
                output += defaultOpts[param];
              }
              else
                output += optsDict[param][data.value];
            } else
              output += data.value;
          }
        } // end param
        return (output)
      },

      _encodeBoolean: function (input) {
        return (input ? '1' : '0');
      },

      _decodeBoolen: function (input) {
        return (input === '1' ? true : false)
      },

      _decodeGeneralSettings: function (input) {
        var outProp = {
          basic: {},
          view: {},
          methylation: {},
          smallrna: {}
        };
        // zoom
        var resultZ = /z([0-9]+)/gi.exec(input);
        if (resultZ != null)
          outProp.basic['highResolutionMode'] = parseInt(resultZ[1]);
        // overview
        var resultO = /o([0-1])/gi.exec(input);
        if (resultO != null)
          outProp.basic['show_overview'] = this._decodeBoolen(resultO[1]);
        // tracklist
        var resultR = /r([0-1])/gi.exec(input);
        if (resultR != null)
          outProp.basic['show_tracklist'] = this._decodeBoolen(resultR[1]);
        // navigation
        var resultN = /n([0-1])/gi.exec(input);
        if (resultN != null)
          outProp.basic['show_nav'] = this._decodeBoolen(resultN[1]);
        // menu
        var resultU = /u([0-1])/gi.exec(input);
        if (resultU != null)
          outProp.basic['show_menu'] = this._decodeBoolen(resultU[1]);
        // labels
        var resultB = /b([0-1])/gi.exec(input);
        if (resultB != null)
          outProp.basic['show_tracklabels'] = this._decodeBoolen(resultB[1]);

        // track padding
        var resultP = /p([0-9]+)/gi.exec(input);
        if (resultP != null)
          outProp.view['trackPadding'] = parseInt(resultP[1]);
        // methylation
        var resultM = /m([0-9]+)/gi.exec(input);
        if (resultM != null) {
          outProp.methylation['CG'] = this._decodeBoolen(resultM[1].substring(0, 1));
          outProp.methylation['CHG'] = this._decodeBoolen(resultM[1].substring(1, 2));
          outProp.methylation['CHH'] = this._decodeBoolen(resultM[1].substring(2, 3));
          if(resultM[1].length > 3){
            outProp.methylation['4mC'] = this._decodeBoolen(resultM[1].substring(3, 4));
            outProp.methylation['5hmC'] = this._decodeBoolen(resultM[1].substring(4, 5));
            outProp.methylation['6mA'] = this._decodeBoolen(resultM[1].substring(5, 6));
          }
        }
        var resultS = /s([0-9]+)/gi.exec(input);
        if (resultS != null) {
          outProp.smallrna['21'] = this._decodeBoolen(resultS[1].substring(0, 1));
          outProp.smallrna['22'] = this._decodeBoolen(resultS[1].substring(1, 2));
          outProp.smallrna['23'] = this._decodeBoolen(resultS[1].substring(2, 3));
          outProp.smallrna['24'] = this._decodeBoolen(resultS[1].substring(3, 4));
          outProp.smallrna['pi'] = this._decodeBoolen(resultS[1].substring(4, 5));
          outProp.smallrna['Others'] = this._decodeBoolen(resultS[1].substring(5, 6));
        }
        return outProp;
      },

      _decodeTrackSettings: function (input, trackLabels) {
        var thisB = this;
        // input and trackLabels are both arrays -- iterate through input
        var out = {};
        array.forEach(input, function (parmStr) {
          var resultT = /^[0-9]+/.exec(parmStr);
          if (resultT === null)
            return;
          var tInt = parseInt(resultT[0]);
          var tLabel = trackLabels[tInt];
          parmStr = parmStr.slice(1);
          out[tLabel] = {};
          var isQuant = null;
          // get quant
          var resultQ = /q([0-1])/gi.exec(parmStr);
          if (resultQ != null) {
            isQuant = thisB._decodeBoolen(resultQ[1]);
            if (isQuant)
              out[tLabel]['style'] = {};
            else
              out[tLabel]['histograms'] = {}
          }
          // get html features
          var isHtml = false;
          var resultV = /v([0-1])/gi.exec(parmStr);
          if (resultV != null) {
            isHtml = thisB._decodeBoolen(resultV[1]);
            if (isHtml) {
              out[tLabel]['type'] = 'ChangeHTMLFeatures';
            }
          }
          // get min
          var resultI = /i(-?[0-9]+(\.[0-9])?)/gi.exec(parmStr);
          //console.log(resultI);
          if (resultI != null) {
            var min = parseFloat(resultI[1]);
            if (isQuant)
              out[tLabel]['min_score'] = min;
            else
              out[tLabel]['histograms']['min'] = min;
          }
          // get max
          var resultX = /x(-?[0-9]+(\.[0-9])?)/gi.exec(parmStr);
          //console.log(resultX);
          if (resultX != null) {
            var max = parseFloat(resultX[1]);
            if (isQuant)
              out[tLabel]['max_score'] = max;
            else
              out[tLabel]['histograms']['max'] = max;
          }
          // get height
          var resultH = /h([0-9]+)/gi.exec(parmStr);
          //console.log(resultH);
          if (resultH != null) {
            var height = parseInt(resultH[1]);
            if (isQuant)
              out[tLabel]['style']['height'] = height;
            else if (isQuant === false) {
              out[tLabel]['maxHeight'] = height;
              out[tLabel]['histograms']['height'] = height;
            } else {
              out[tLabel]['maxHeight'] = height;
            }
          }
          // get ypos
          var resultY = /y([0-3])/gi.exec(parmStr);
          //console.log(resultY);
          if (resultY != null) {
            var locList = ['none', 'center', 'left', 'right'];
            var yposI = parseInt(resultY[1]);
            out[tLabel]['yScalePosition'] = locList[yposI];
          }
          // get mode
          var resultD = /d([0-2])/gi.exec(parmStr);
          //console.log(resultY);
          if (resultD != null) {
            var modeList = ['normal', 'compact', 'collapsed'];
            var modeI = parseInt(resultD[1]);
            out[tLabel]['displayMode'] = modeList[modeI];
          }
          // get style
          var resultF = /f([0-2])/gi.exec(parmStr);
          //console.log(resultY);
          if (resultF != null) {
            var styleList = ['default', 'features', 'histograms'];
            var styleI = parseInt(resultF[1]);
            out[tLabel]['displayStyle'] = styleList[styleI];
          }
        });
        return out;
      }

    }
    return EncodeDecodeUtil;
  });
